app/warden/magic_link_token.rb

```diff
@@ -0,0 +1,15 @@
+class MagicLinkToken < Warden::Strategies::Base
+  def valid?
+    params[:token].present?
+  end
+
+  def authenticate!
+    candidate = MagicLinkFlow.find_authenticated_candidate(raw_token: params[:token])
+
+    if candidate
+      success!(candidate)
+    else
+      fail!('Invalid magic link')
```

Comment:
Is this string required? Do we test the unhappy path?

Reply:
Nope, the string isn't required. We've removed it. In regards to testing the unhappy path, we test if an invalid token is provided:

https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/0379ec8d332b7e5ad23f6dd10787423de8505171/spec/system/candidate_signing_up_spec.rb#L36-L42